src/WaitosaurBase.sol

```diff
@@ -0,0 +1,162 @@
+    bytes32 internal constant STATE_STORAGE_SLOT =
+        0x3558e4a352cfd953dc9839fb713b54b0ad22b91e762cb85051783d80d3379c00;
+    /// @dev keccak256(abi.encode(uint256(keccak256("maxbtc.waitosaur.base.roles")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 internal constant ROLES_STORAGE_SLOT =
+        0xf8491acbe0d7b7ea9ed2881253b95cb844f2ef763d00e27dc8068974c606e200;
+
+    function _getState() internal pure returns (WaitosaurState storage s) {
+        assembly {
+            s.slot := STATE_STORAGE_SLOT
+        }
+    }
+
+    function getState() external pure returns (WaitosaurState memory) {
+        WaitosaurState storage state = _getState();
+        return state;
+    }
+
+    function _getRoles() internal pure returns (WaitosaurAccess storage r) {
+        assembly {
+            r.slot := ROLES_STORAGE_SLOT
+        }
+    }
+
+    function getRoles() public pure returns (WaitosaurAccess memory) {
+        WaitosaurAccess storage r = _getRoles();
+        return r;
+    }
+
+    function __WaitosaurBase_init(
+        address owner_,
+        address locker_,
+        address unlocker_
+    ) internal onlyInitializing {
+        __Ownable_init(owner_);
+        __Ownable2Step_init();
+        __UUPSUpgradeable_init();
+
+        _setRoles(locker_, unlocker_);
+        _clearLock(_getState());
+    }
+
+    function _setRoles(address locker, address unlocker) internal {
+        WaitosaurAccess storage r = _getRoles();
+        require(
+            locker != address(0) && unlocker != address(0),
+            InvalidRolesAddresses()
+        );
+
+        r.locker = locker;
+        r.unlocker = unlocker;
+
+        emit RolesUpdated(r);
+    }
+
+    function updateRoles(
+        address newLocker,
+        address newUnlocker
+    ) public onlyOwner {
+        _setRoles(newLocker, newUnlocker);
+    }
+
+    function _lockBase(
+        uint256 amount
+    ) internal returns (WaitosaurState storage state) {
+        if (amount == 0) revert AmountZero();
+        state = _getState();
+        if (state.lockedAmount != 0) revert AlreadyLocked();
+        state.lockedAmount = amount;
+        state.lastLocked = block.timestamp;
+
+        emit Locked(amount);
+    }
+
+    function lock(uint256 amount) public {
+        WaitosaurAccess storage roles = _getRoles();
+        if (_msgSender() != roles.locker && _msgSender() != owner()) {
+            revert Unauthorized();
+        }
+        _lockBase(amount);
+    }
+
+    /// @dev Only unlocker or owner is allowed.
+    function unlock() external {
+        WaitosaurAccess storage roles = _getRoles();
+        if (_msgSender() != roles.unlocker && _msgSender() != owner())
+            revert Unauthorized();
+        WaitosaurState storage state = _getState();
+        _ensureLocked(state);
+        _unlock(state);
+        _clearLock(state);
+        emit Unlocked();
+    }
+
+    function _unlock(WaitosaurState storage state) internal virtual {}
```

Comment:
I think this function shouldn't accept any arguments. It reduces flexibility for future child classes which will need to override it and wouldn't have to use `WaitosaurState`.